tests/misc_env/homogenous_osd_cluster.py

```diff
@@ -0,0 +1,156 @@
+    a = grouped_data
+    size_counts = {}
+    size_suffix = {}
+    # Loop through the data to count devices for each size
+    for host, sizes_and_devices in a.items():
+        for size_str, devices in sizes_and_devices.items():
+            size_value = float(size_str.split()[0])
+            size_counts[size_value] = size_counts.get(size_value, 0) + len(devices)
+            size_suffix[size_value] = size_str.split()[1]
+    # Find the size with the maximum device count
+    max_size_value = max(size_counts, key=size_counts.get)
+    # Convert the maximum available size to a string with the appropriate unit
+    desired_size = f"{max_size_value} {size_suffix[max_size_value]}"
+    # Get the maximum device count for the desired size
+    max_device_count = size_counts[max_size_value]
+    # Initialize a variable to keep track of the minimum node count
+    min_disk_count = float("inf")
+    # Loop through the data to find the minimum node count for the desired size
+    for host, sizes_and_devices in a.items():
+        if desired_size in sizes_and_devices:
+            size_device_count = len(sizes_and_devices[desired_size])
+            min_disk_count = min(min_disk_count, size_device_count)
+    # Print the desired size, the maximum number of devices, and the maximum number of nodes for that size
+    log.info(
+        f"The desired size based on max available devices across all nodes is: {desired_size}"
+    )
+    log.info(f"The maximum number of devices for {desired_size} is: {max_device_count}")
+    if min_disk_count == float("inf"):
+        log.info("No nodes have the desired size.")
+    else:
+        log.info(
+            f"Maximum disks of size {desired_size} that can be applied across all nodes: {min_disk_count}"
+        )
+    return grouped_data, desired_size, min_disk_count
+
+
+def deploy_osd(node: CephNode, osn, disk_list: List) -> None:
+    """
+    Deploy OSD on specific device
+    Args:
+        node:   Installer node
+        osn:    OSD node name
+        osd_node : node_ip of node where osd is to be deployed
+        disk_list:   List of devices to deploy OSD
+    Returns:
+        None
+    Raises:
+        CommandFailed
+    """
+    disk_list_updated = []
+    for i in range(len(disk_list)):
+        try:
+            print(f"cephadm shell -- ceph orch daemon add osd {osn}:{disk_list[i]}")
+            node.exec_command(
+                sudo=True,
+                cmd=f"cephadm shell -- ceph orch daemon add osd {osn}:{disk_list[i]}",
+            )
+            disk_list_updated.append(disk_list[i])
+        except CommandFailed:
+            log.info(f"Could not deploy OSD on {disk_list[i]}")
+    expected_osd_count = len(disk_list_updated)
+    validate_osd(node, osn, expected_osd_count)
+
+
+@retry(AssertionError, tries=15, delay=10)
+def validate_osd(node: CephNode, osn, expected_osds):
+    out, _ = node.exec_command(
+        sudo=True, cmd="cephadm shell -- ceph orch ps --format json-pretty"
```

Comment:
Can we replace above command with 
cephadm shell -- ceph orch ps dmn["hostname"] dmn["hostname"] --format json-pretty

This will direclty give the osd under that hostname. we can just get with len(out)
Sample out put
[root@ceph-nfs-reg-q0a62t-node8 ~]# ceph orch ps ceph-nfs-reg-q0a62t-node4 --daemon_type=osd
NAME   HOST                       PORTS  STATUS         REFRESHED  AGE  MEM USE  MEM LIM  VERSION          IMAGE ID      CONTAINER ID  
osd.0  ceph-nfs-reg-q0a62t-node4         running (13h)          -  15h     299M    4096M  18.2.0-39.el9cp  a29cdf5c35dd  236f52508b0d  
osd.3  ceph-nfs-reg-q0a62t-node4         running (14h)          -  15h     822M    4096M  18.2.0-39.el9cp  a29cdf5c35dd  ff90b59fddd3  
osd.6  ceph-nfs-reg-q0a62t-node4         running (15h)          -  15h     760M    4096M  18.2.0-39.el9cp  a29cdf5c35dd  b843542953d4  
osd.9  ceph-nfs-reg-q0a62t-node4         running (14h)          -  15h     894M    4096M  18.2.0-39.el9cp  a29cdf5c35dd  f11d1502f2b1  

[root@ceph-nfs-reg-q0a62t-node8 ~]# ceph orch ps ceph-nfs-reg-q0a62t-node4 --daemon_type=osd --format json

[{"container_id": "236f52508b0d", "container_image_digests": ["registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:283bb3dfc9b0c819343091eafb4b17f46030f2fedb78ad90dd001f809383e862"], "container_image_id": "a29cdf5c35dda572e65e7a8abfbde05fdbdadd6923deb05b15e0eefbcef77dd9", "container_image_name": "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "cpu_percentage": "0.28%", "created": "2023-09-19T17:24:27.281757Z", "daemon_id": "0", "daemon_name": "osd.0", "daemon_type": "osd", "events": ["2023-09-19T17:24:27.314241Z daemon:osd.0 [INFO] \"Deployed osd.0 on host 'ceph-nfs-reg-q0a62t-node4'\""], "hostname": "ceph-nfs-reg-q0a62t-node4", "is_active": false, "memory_request": 4294967296, "memory_usage": 313838796, "ports": [], "service_name": "osd.all-available-devices", "started": "2023-09-19T19:23:25.281221Z", "status": 1, "status_desc": "running", "version": "18.2.0-39.el9cp"}, {"container_id": "ff90b59fddd3", "container_image_digests": ["registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:283bb3dfc9b0c819343091eafb4b17f46030f2fedb78ad90dd001f809383e862"], "container_image_id": "a29cdf5c35dda572e65e7a8abfbde05fdbdadd6923deb05b15e0eefbcef77dd9", "container_image_name": "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "cpu_percentage": "0.54%", "created": "2023-09-19T17:24:29.772771Z", "daemon_id": "3", "daemon_name": "osd.3", "daemon_type": "osd", "events": ["2023-09-19T17:24:29.805249Z daemon:osd.3 [INFO] \"Deployed osd.3 on host 'ceph-nfs-reg-q0a62t-node4'\""], "hostname": "ceph-nfs-reg-q0a62t-node4", "is_active": false, "memory_request": 4294967296, "memory_usage": 861929472, "ports": [], "service_name": "osd.all-available-devices", "started": "2023-09-19T18:23:20.248775Z", "status": 1, "status_desc": "running", "version": "18.2.0-39.el9cp"}, {"container_id": "b843542953d4", "container_image_digests": ["registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:283bb3dfc9b0c819343091eafb4b17f46030f2fedb78ad90dd001f809383e862"], "container_image_id": "a29cdf5c35dda572e65e7a8abfbde05fdbdadd6923deb05b15e0eefbcef77dd9", "container_image_name": "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "cpu_percentage": "0.99%", "created": "2023-09-19T17:24:32.077784Z", "daemon_id": "6", "daemon_name": "osd.6", "daemon_type": "osd", "events": ["2023-09-19T17:24:32.109796Z daemon:osd.6 [INFO] \"Deployed osd.6 on host 'ceph-nfs-reg-q0a62t-node4'\""], "hostname": "ceph-nfs-reg-q0a62t-node4", "is_active": false, "memory_request": 4294967296, "memory_usage": 797756620, "ports": [], "service_name": "osd.all-available-devices", "started": "2023-09-19T17:24:32.028988Z", "status": 1, "status_desc": "running", "version": "18.2.0-39.el9cp"}, {"container_id": "f11d1502f2b1", "container_image_digests": ["registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:283bb3dfc9b0c819343091eafb4b17f46030f2fedb78ad90dd001f809383e862"], "container_image_id": "a29cdf5c35dda572e65e7a8abfbde05fdbdadd6923deb05b15e0eefbcef77dd9", "container_image_name": "registry-proxy.engineering.redhat.com/rh-osbs/rhceph@sha256:2495093358eab77c64279816bb57c20fe8604118d925c3e097f53d98e729bda8", "cpu_percentage": "0.55%", "created": "2023-09-19T17:24:34.412797Z", "daemon_id": "9", "daemon_name": "osd.9", "daemon_type": "osd", "events": ["2023-09-19T17:24:34.445094Z daemon:osd.9 [INFO] \"Deployed osd.9 on host 'ceph-nfs-reg-q0a62t-node4'\""], "hostname": "ceph-nfs-reg-q0a62t-node4", "is_active": false, "memory_request": 4294967296, "memory_usage": 937846374, "ports": [], "service_name": "osd.all-available-devices", "started": "2023-09-19T18:17:26.632406Z", "status": 1, "status_desc": "running", "version": "18.2.0-39.el9cp"}]

With this, we can avoid all the if and for loops around